lang/stringx/stringx.go

```diff
@@ -0,0 +1,291 @@
+		return s
+	}
+	return strings.ReplaceAll(s, rmStr, "")
+}
+
+// Rotate rotates(circular shift) a string of shift characters.
+func Rotate(s string, shift int) string {
+	if shift == 0 {
+		return s
+	}
+	sLen := len(s)
+	if sLen == 0 {
+		return s
+	}
+
+	shiftMod := shift % sLen
+	if shiftMod == 0 {
+		return s
+	}
+
+	offset := -(shiftMod)
+	sb := strings.Builder{}
+	sb.Grow(sLen)
+	_, _ = sb.WriteString(SubStart(s, offset))
+	_, _ = sb.WriteString(Sub(s, 0, offset))
+	return sb.String()
+}
+
+// Sub returns substring from specified string avoiding panics with index start and end.
+// start, end are based on unicode(utf8) count
+func Sub(s string, start, end int) string {
+	return sub(s, start, end)
+}
+
+// SubStart returns substring from specified string avoiding panics with start.
+// start, end are based on unicode(utf8) count
+func SubStart(s string, start int) string {
+	return sub(s, start, math.MaxInt64)
+}
+
+func sub(s string, start, end int) string {
+	if s == "" {
+		return ""
+	}
+
+	unicodeLen := Utf8Len(s)
+	// end
+	if end < 0 {
+		end += unicodeLen
+	}
+	if end > unicodeLen {
+		end = unicodeLen
+	}
+	// start
+	if start < 0 {
+		start += unicodeLen
+	}
+	if start > end {
+		return ""
+	}
+
+	// start <= end
+	if start < 0 {
+		start = 0
+	}
+	if end < 0 {
+		end = 0
+	}
+	if start == 0 && end == unicodeLen {
+		return s
+	}
+
+	sb := strings.Builder{}
+	sb.Grow(end - start)
+	runeIndex := 0
+	for _, v := range s {
+		if runeIndex >= end {
+			break
+		}
+		if runeIndex >= start {
+			sb.WriteRune(v)
+		}
+		runeIndex++
+	}
+	return sb.String()
+}
+
+// Reverse reverses a string
+func Reverse(s string) string {
+	result, _ := ReverseE(s)
+	return result
+}
+
+// ReverseE reverses a string with error status returned
+func ReverseE(s string) (string, error) {
+	if s == "" {
+		return s, nil
+	}
+	src := hack.StringToBytes(s)
+	dst := make([]byte, len(s))
+	srcIndex := len(s)
+	dstIndex := 0
+	for srcIndex > 0 {
+		r, n := utf8.DecodeLastRune(src[:srcIndex])
+		if r == utf8.RuneError {
+			return hack.BytesToString(dst), ErrDecodeRune
+		}
+		utf8.EncodeRune(dst[dstIndex:], r)
+		srcIndex -= n
+		dstIndex += n
+	}
+	return hack.BytesToString(dst), nil
+}
+
+// Shuffle  shuffles runes in a string and returns.
+func Shuffle(s string) string {
+	if s == "" {
+		return s
+	}
+	runes := []rune(s)
+	index := 0
+	for i := len(runes) - 1; i > 0; i-- {
+		index = rand.Intn(i + 1)
+		if i != index {
+			runes[i], runes[index] = runes[index], runes[i]
+		}
+	}
+	return string(runes)
+}
+
+// ContainsAnySubstrings returns whether s contains any of substring in slice.
+func ContainsAnySubstrings(s string, subs []string) bool {
+	for _, v := range subs {
+		if strings.Contains(s, v) {
+			return true
+		}
+	}
+	return false
+}
+
+// Utf8Len returns str's utf8 rune length.
+// Erroneous and short encodings are treated as single runes of width 1 byte.
+func Utf8Len(s string) int {
+	return utf8.RuneCountInString(s)
```

Comment:
same as above, we can use utf8 directly.

Reply:
> same as above, we can use utf8 directly.

RuneCountInString is not a straight name,
and it's not a nonessential feature, i'll disable it temporarily.  When if there is actual need from feedback in future,  We can start the discussion again.